source/sink/service_impl.cc

```diff
@@ -0,0 +1,132 @@
+  return grpc_status;
+}
+
+grpc::Status SinkServiceImpl::SinkRequestStream(
+    grpc::ServerContext*,
+    grpc::ServerReaderWriter<nighthawk::SinkResponse, nighthawk::SinkRequest>* stream) {
+  nighthawk::SinkRequest request;
+  while (stream->Read(&request)) {
+    ENVOY_LOG(trace, "Inbound SinkRequest {}", request.DebugString());
+    absl::StatusOr<std::vector<nighthawk::client::ExecutionResponse>> execution_responses =
+        sink_->LoadExecutionResult(request.execution_id());
+    if (!execution_responses.status().ok()) {
+      return abslStatusToGrpcStatus(execution_responses.status());
+    }
+    absl::StatusOr<nighthawk::client::ExecutionResponse> response =
+        mergeExecutionResponses(request.execution_id(), *execution_responses);
+    if (!response.status().ok()) {
+      return abslStatusToGrpcStatus(response.status());
+    }
+    nighthawk::SinkResponse sink_response;
+    *(sink_response.mutable_execution_response()) = *response;
+    if (!stream->Write(sink_response)) {
+      return abslStatusToGrpcStatus(
+          absl::Status(absl::StatusCode::kInternal, "Failure writing response to stream."));
+    }
+  }
+  return abslStatusToGrpcStatus(absl::OkStatus());
+}
+
+absl::Status mergeOutput(const nighthawk::client::Output& input_to_merge,
+                         nighthawk::client::Output& merge_target) {
+  if (!merge_target.has_options()) {
+    // If no options are set, that means this is the first part of the merge.
+    // Set some properties that shouldbe equal amongst all Output instances.
+    *merge_target.mutable_options() = input_to_merge.options();
+    *merge_target.mutable_timestamp() = input_to_merge.timestamp();
+    *merge_target.mutable_version() = input_to_merge.version();
+  } else {
+    // Options used should not diverge for a executions under a single execution id.
+    // Versions probably shouldn't either. We sanity check these things here, and
+    // report on error when we detect any mismatch.
+    if (!MessageDifferencer::Equivalent(input_to_merge.options(), merge_target.options())) {
+      return absl::Status(absl::StatusCode::kInternal,
+                          fmt::format("Options divergence detected: {} vs {}.",
+                                      merge_target.options().DebugString(),
+                                      input_to_merge.options().DebugString()));
+    }
+    if (!MessageDifferencer::Equivalent(input_to_merge.version(), merge_target.version())) {
+      return absl::Status(absl::StatusCode::kInternal,
+                          fmt::format("Version divergence detected: {} vs {}.",
+                                      merge_target.version().DebugString(),
+                                      input_to_merge.version().DebugString()));
+    }
+  }
+  // Append all input results into our own results.
+  for (const nighthawk::client::Result& result : input_to_merge.results()) {
+    merge_target.add_results()->MergeFrom(result);
```

Comment:
to my knowledge we don't have code anywhere that alters result names based on which nighthawk they came from.  As a result, I believe that if you're using three nighthawks, you'd wind up with three results with the name "global" and if those nighthawks each had two workers, you'd also have three of each "worker1" and "worker2" (or however they're named).

Reply:
Oh, good catch, this looses some information unintentionally. Thank you for catching that!

I think we could improve here by having a new field in `nighthawk::client::Result`, where we can tag the originating (listening address of the) associated load gen service, so we can trace back where the result came from.. As it stands, we do know due to the way that this is structured that the results are distinct, but we do not know which service was responsible for what results. I'll propose something to address this.